pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,460 @@
+		desired:    obj,
+	}
+
+	// Get compute GCP client
+	if location == "global" {
+		gcpClient, err := m.globalForwardingRulesClient(ctx)
+		if err != nil {
+			return nil, err
+		}
+		forwardingRuleAdapter.globalForwardingRulesClient = gcpClient
+	} else {
+		gcpClient, err := m.forwardingRulesClient(ctx)
+		if err != nil {
+			return nil, err
+		}
+		forwardingRuleAdapter.forwardingRulesClient = gcpClient
+	}
+	return forwardingRuleAdapter, nil
+}
+
+func (m *forwardingRuleModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+func (a *forwardingRuleAdapter) Find(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting ComputeForwardingRule", "name", a.resourceID)
+
+	var err error
+	forwardingRule := &computepb.ForwardingRule{}
+	if a.Location == "global" {
+		req := &computepb.GetGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		forwardingRule, err = a.globalForwardingRulesClient.Get(ctx, req)
+	} else {
+		req := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		forwardingRule, err = a.forwardingRulesClient.Get(ctx, req)
+	}
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+	}
+	a.actual = forwardingRule
+	return true, nil
+}
+
+func (a *forwardingRuleAdapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	if a.projectID == "" {
+		return fmt.Errorf("project is empty")
+	}
+	if a.resourceID == "" {
+		return fmt.Errorf("resourceID is empty")
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating ComputeForwardingRule", "name", a.resourceID)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+
+	forwardingRule := ComputeForwardingRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	forwardingRule.Name = direct.LazyPtr(a.resourceID)
+	forwardingRule.Labels = desired.Labels
+	// TODO(yuhou): handle default spec values
+	if forwardingRule.LoadBalancingScheme == nil {
+		forwardingRule.LoadBalancingScheme = direct.LazyPtr("EXTERNAL")
+	}
+
+	var err error
+	op := &gcp.Operation{}
+	created := &computepb.ForwardingRule{}
+	if a.Location == "global" {
+		req := &computepb.InsertGlobalForwardingRuleRequest{
+			ForwardingRuleResource: forwardingRule,
+			Project:                a.projectID,
+		}
+		getReq := &computepb.GetGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		op, err = a.globalForwardingRulesClient.Insert(ctx, req)
+		if err != nil {
+			return fmt.Errorf("creating ComputeForwardingRule %s failed: %w", a.resourceID, err)
```

Comment:
In other controllers, we've used `fullyQualifiedName()` instead of just the name, to give a more complete error message